### DISCUSSION TITLE
Why doesn't ui.select update the value after changing its options?

### DISCUSSION THREAD

**Discussion by spehj** [zauberzeug/nicegui#508] [Q&A] [answered]
I'm trying to use `select` element for selecting a serial port that is found after searching. When I run this example for the first time, `select` element doesn't update _value_ to the first element in the list. Also, the first element in the dropdown list can't be selected for the first time. I have to select second or third one and then I can select first one again. What could be the problem?

Here is a minimum example:
```py
import time
from nicegui import app, ui
import asyncio
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor

pool = ThreadPoolExecutor()

def search_ports():
        # Some blocking function searching for available serial ports.
        time.sleep(3)
        return ["COM4", "COM6", "COM8"]
  
async def start_ports_search():
        # Function to start searching and update UI on result
        ui.notify("Searching available Ports")
        loop = asyncio.get_running_loop()
        result = await loop.run_in_executor(pool, search_ports)
        ui.notify(result)
        select_ports.options = result
        select_ports.value = result[0]
        select_ports.update()

with ui.row():
        ui.button("SEARCH PORTS", on_click=start_ports_search)
        select_ports = ui.select([]).props('outlined label="Select Port"').classes("w-32")

app.on_shutdown(pool.shutdown)
ui.run()

```

**Comment by falkoschindler** [comment, marked as answer]
As far as I understand your problem is as follows:

```py
select = ui.select([])

def fill_options():
    select.options = ['a', 'b', 'c']
    select.value = 'a'
    select.update()

ui.button('Fill options', on_click=fill_options)
```

The button should populate the selection element with three options and set its value to the first one. But the displayed value remains blank. And when trying to manually select the first value, it doesn't change.

The reason is that some internal data structures need to get updated using `select.update()` after changing the `options` field. When setting the `value`, its index is computed internally and sent to the client. But if the `value` is set _before_ triggering the update, the index is invalid and causes the strange behavior you are describing.

So adding a `select.update()` before setting the new `value = 'a'` solves the problem.

**Reply by spehj** [reply]
Thank you! This solves the problem.
